node/core/provisioner/src/disputes/prioritized_selection/mod.rs

```diff
@@ -0,0 +1,445 @@
+}
+
+impl PartitionedDisputes {
+	fn new() -> PartitionedDisputes {
+		Default::default()
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.inactive_unknown_onchain
+			.into_iter()
+			.chain(self.active_unknown_onchain.into_iter())
+			.chain(self.active_unconcluded_onchain.into_iter())
+			.chain(self.active_concluded_onchain.into_iter())
+		// inactive_known_onchain is dropped on purpose
+	}
+}
+
+fn secs_since_epoch() -> Timestamp {
+	match SystemTime::now().duration_since(UNIX_EPOCH) {
+		Ok(d) => d.as_secs(),
+		Err(e) => {
+			gum::warn!(
+				target: LOG_TARGET,
+				err = ?e,
+				"Error getting system time."
+			);
+			0
+		},
+	}
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(SessionIndex, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+
+	// Drop any duplicates
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	// Split ACTIVE from CONCLUDED disputes
+	let (active, inactive): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| !dispute_is_inactive(status, &secs_since_epoch()));
+
+	// Split ACTIVE in three groups...
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => {
+				// Check if there are enough onchain votes for or against to conclude the dispute
+				let supermajority = supermajority_threshold(d.validators_for.len());
+				if d.validators_for.count_ones() >= supermajority ||
+					d.validators_against.count_ones() >= supermajority
+				{
+					partitioned.active_concluded_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.active_unconcluded_onchain.push((session_index, candidate_hash));
+				}
+			},
+			None => partitioned.active_unknown_onchain.push((session_index, candidate_hash)),
+		};
+	}
+
+	// ... and INACTIVE in two more
+	for (session_index, candidate_hash, _) in inactive {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(_) => partitioned.inactive_known_onchain.push((session_index, candidate_hash)),
+			None => partitioned.inactive_unknown_onchain.push((session_index, candidate_hash)),
+		}
+	}
+
+	partitioned
+}
+
+// Helper trait to obtain the value of vote for `InvalidDisputeStatementKind` and `ValidDisputeStatementKind`.
+// The alternative was to pass a bool to `fn is_vote_worth_to_keep` explicitly but it's pointless as the value is already 'encoded' in the type.
+trait VoteType {
+	fn vote_value() -> bool;
```

Comment:
if this was named like `is_valid()` I think code would be easier to grasp. Another alternative would be to pass [`DisputeStatement`](https://github.com/paritytech/polkadot/blob/b4cf84ddf8cf5c7ecfaf2cb906f502c658fe7694/primitives/src/v2/mod.rs#L1214) and do a pattern match.